src/sentry/api/bases/monitor.py

```diff
@@ -57,23 +60,56 @@ class MonitorEndpoint(Endpoint):
     def respond_invalid() -> Response:
         return Response(status=status.HTTP_400_BAD_REQUEST, data={"details": "Invalid monitor"})
 
-    def convert_args(self, request: Request, monitor_id: str, *args, **kwargs):
-        try:
-            UUID(monitor_id)
-        except ValueError:
-            raise ParameterValidationError("Invalid monitor UUID")
+    def convert_args(
+        self,
+        request: Request,
+        monitor_id: str,
+        organization_slug: str | None = None,
+        *args,
+        **kwargs,
+    ):
+        organization = None
+        monitor = None
 
-        try:
-            monitor = Monitor.objects.get(guid=monitor_id)
-        except Monitor.DoesNotExist:
-            raise ResourceDoesNotExist
+        # The only monitor endpoints that do not have the org slug in their
+        # parameters are the GUID-style checkin endpoints
+        if organization_slug:
+            try:
+                organization = Organization.objects.get_from_cache(slug=organization_slug)
+                # Try lookup by slug first. This requires organization context since
+                # slugs are unique only to the organization
+                monitor = Monitor.objects.get(organization_id=organization.id, slug=monitor_id)
+            except (Organization.DoesNotExist, Monitor.DoesNotExist):
+                pass
+
+        # Try lookup by GUID
+        if not monitor:
+            # Validate GUIDs
+            try:
+                UUID(monitor_id)
+            except ValueError:
+                # This error is a bit confusing, because this may also mean
+                # that we've failed to lookup their monitor by slug.
+                raise ParameterValidationError("Invalid monitor UUID")
+            # When looking up by guid we don't include the org conditional
+            # (since GUID lookup allows orgless routes), we will validate
+            # permissions later in this function
+            try:
+                monitor = Monitor.objects.get(guid=monitor_id)
+            except Monitor.DoesNotExist:
+                raise ResourceDoesNotExist
 
         project = Project.objects.get_from_cache(id=monitor.project_id)
         if project.status != ProjectStatus.VISIBLE:
             raise ResourceDoesNotExist
 
         if hasattr(request.auth, "project_id") and project.id != request.auth.project_id:
-            raise InvalidAuthProject
+            raise ResourceDoesNotExist
+
+        # When looking up via GUID we do not check the organiation slug,
+        # validate that the slug matches the org of the monitors project
+        if organization_slug and project.organization.slug != organization_slug:
```

Comment:
Does the project already have the organization eager loaded, or are we hitting the DB again here?

Reply:
I noticed this too. It's already been like this for a while. I'll update.

Reply:
In a follow up that is